### PR TITLE
Non-pure function detected.

### PR DESCRIPTION
**Pull request by TaMaN2031A**
### Description

I'm working on detecting the impure functions in the code, mentioning them in the header file of it, and trying to enhance them if possible.

### CODE REVIEW

File: moveit_core/collision_detection/src/collision_tools.cpp [hunk 1]
```diff
@@ -39,7 +39,7 @@
 #include <rclcpp/time.hpp>
 #include <tf2_eigen/tf2_eigen.hpp>
 
-namespace collision_detection
+namespace collision_detection // This namespace is perfectly pure, nothing to be modified actually, perfect.
```
Comment (AndyZe) [inline moveit_core/collision_detection/src/collision_tools.cpp:42]: Can you move this comment to the header file, "collision_tools.h"? I suggest this, I think it reads more easily:

```
namespace collision_detection
{
// Perfectly "pure" helper functions
```

File: moveit_core/collision_detection/src/world_diff.cpp [hunk 2]
```diff
@@ -96,7 +96,7 @@ void WorldDiff::reset(const WorldPtr& world)
 void WorldDiff::setWorld(const WorldPtr& world)
 {
   WorldPtr old_world = world_.lock();
-  if (old_world)
+  if (old_world) // Variable world_ is encapsulated in the class, this is important. But not being a parameter it causes impure function
```
Comment (AndyZe) [inline moveit_core/collision_detection/src/world_diff.cpp:99]: Suggestion... for some reason, github is not letting me make a suggestion the normal way.

```
// Warning: function output depends on the class member `world_`. This makes it an "impure function" 
if (old_world)
```

File: moveit_core/dynamics_solver/src/dynamics_solver.cpp [hunk 3]
```diff
@@ -281,7 +281,7 @@ bool DynamicsSolver::getMaxPayload(const std::vector<double>& joint_angles, doub
 }
 
 bool DynamicsSolver::getPayloadTorques(const std::vector<double>& joint_angles, double payload,
-                                       std::vector<double>& joint_torques) const
+                                       std::vector<double>& joint_torques) const // NonPure, and again they depend on the joint_model_group_ boolean variable
```
Comment (AndyZe) [inline moveit_core/dynamics_solver/src/dynamics_solver.cpp:284]: I think just one warning in the header file would be enough, please. Labeling every singe function is excessive IMO.

Can you please delete all comments here and put a warning like this in `dynamics_solver.h`?

// Warning: many of these functions are "impure", meaning their output depends on the state of member variables in a non-transparent way. For example, `getMaxPayload()` depends on `joint_model_group_`.

File: moveit_ros/perception/semantic_world/src/semantic_world.cpp [hunk 4]
```diff
@@ -146,7 +146,7 @@ bool SemanticWorld::addTablesToCollisionWorld()
       triangles[i3] = j + 1;
     }
 
-    shapes::Shape* table_shape = shapes::createMeshFromVertices(vertices, triangles);
+    shapes::Shape* table_shape = shapes::createMeshFromVertices(vertices, triangles); // We could move this part to a new function to keep the function pure
```
Comment (AndyZe) [inline moveit_ros/perception/semantic_world/src/semantic_world.cpp:149]: I rather would see you simply open a PR to make this change :+1:

```suggestion
    shapes::Shape* table_shape = shapes::createMeshFromVertices(vertices, triangles);
```

**Review by AndyZe — COMMENTED** [review]
Thanks for pointing this out! It's an important concept.

File: moveit_core/dynamics_solver/src/dynamics_solver.cpp (hunk 3 above)
Comment (TaMaN2031A) [reply moveit_core/dynamics_solver/src/dynamics_solver.cpp:284]: OK, sir, I'll work on modifying them today.

File: moveit_core/collision_detection/src/world_diff.cpp (hunk 2 above)
Comment (TaMaN2031A) [reply moveit_core/collision_detection/src/world_diff.cpp:99]: > Suggestion... for some reason, github is not letting me make a suggestion the normal way.
> 
> ```
> // Warning: function output depends on the class member `world_`. This makes it an "impure function" 
> if (old_world)
> ```

Thank you for your time! Shall I write all the comments in the future in this way?

File: moveit_ros/perception/semantic_world/src/semantic_world.cpp (hunk 4 above)
Comment (TaMaN2031A) [reply moveit_ros/perception/semantic_world/src/semantic_world.cpp:149]: OK, sir, I'll work on it this week inshallah.

Comment (TaMaN2031A) [reply moveit_ros/perception/semantic_world/src/semantic_world.cpp]: Thank you for your time. This is not possible for me at the moment actually as I don't have a whole idea about the code, I'll only work in small pieces with the help of my simple knowledge of functional programming. 
Soon this will be possible inshaAllah.

**Comment by AndyZe** [comment]
Looks good to me besides one small comment. We need to wait for a second reviewer now. Thanks for your time.

**Review by AndyZe — APPROVED** [review]

**Comment by AndyZe** [comment]
Yeah, unfortunately it is  :(  It's only been a week; that's not too bad.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_tools.h [hunk 5]
```diff
@@ -48,6 +48,7 @@
 
 namespace collision_detection
 {
+// Perfectly "pure" helper functions
```
Comment (henningkayser) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_tools.h:51]: why state this here? It doesn't document anything noteworthy about this free function.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_env.h [hunk 6]
```diff
@@ -47,7 +47,10 @@ namespace collision_detection
 {
 MOVEIT_CLASS_FORWARD(CollisionEnv);  // Defines CollisionEnvPtr, ConstPtr, WeakPtr... etc
 
-/** \brief Provides the interface to the individual collision checking libraries. */
+/** \brief Provides the interface to the individual collision checking libraries.
+ * Some of these functions are impure, they depend on some shared variable .
+ *  The world_ variable for example is controlling the return value though it
```
Comment (henningkayser) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_env.h:52]: what return value? world_ can be set using the class API

File: moveit_core/collision_detection/include/moveit/collision_detection/test_collision_common_panda.h [hunk 7]
```diff
@@ -366,3 +366,4 @@ REGISTER_TYPED_TEST_SUITE_P(CollisionDetectorPandaTest, InitOK, DefaultNotInColl
 REGISTER_TYPED_TEST_SUITE_P(DistanceCheckPandaTest, DistanceSingle);
 
 REGISTER_TYPED_TEST_SUITE_P(DistanceFullPandaTest, DistancePoints);
+// namespace collision_detection
```
Comment (henningkayser) [inline moveit_core/collision_detection/include/moveit/collision_detection/test_collision_common_panda.h:369]: out of place + unrelated change

File: moveit_core/collision_detection/include/moveit/collision_detection/world_diff.h [hunk 8]
```diff
@@ -59,8 +59,9 @@ class WorldDiff
   ~WorldDiff();
 
   /** \brief Set which world to record.  Records all objects in old world (if
-   * any) as DESTROYED and all objects in new world as CREATED and ADD_SHAPE
-   * */
+   * any) as DESTROYED and all objects in new world as CREATED and ADD_SHAPE.
+   * Warning: function output depends on the class member `world_`. This makes it an "impure function"
```
Comment (henningkayser) [inline moveit_core/collision_detection/include/moveit/collision_detection/world_diff.h:63]: This is the setter for world_. What other side effects do you expect to happen here?

File: moveit_core/dynamics_solver/include/moveit/dynamics_solver/dynamics_solver.h [hunk 9]
```diff
@@ -48,6 +48,8 @@
 /** \brief This namespace includes the dynamics_solver library */
 namespace dynamics_solver
 {
+// Warning: many of these functions are "impure", meaning their output depends on the state of member variables in a
```
Comment (henningkayser) [inline moveit_core/dynamics_solver/include/moveit/dynamics_solver/dynamics_solver.h:51]: seems completely out of place to me

File: moveit_ros/warehouse/include/moveit/warehouse/planning_scene_world_storage.h [hunk 10]
```diff
@@ -41,6 +41,8 @@
 
 namespace moveit_warehouse
 {
+// Warning: some of these functions are "impure", meaning their output depends on the state of member variables in a
```
Comment (henningkayser) [inline moveit_ros/warehouse/include/moveit/warehouse/planning_scene_world_storage.h:44]: Out of place, and the comment is also meaningless in this context.

**Review by henningkayser — CHANGES_REQUESTED** [review]
I don't see the use of any of these comments.  non-const class member functions will obviously depend on the internal state of the class.  Most of these comments are not even true. Just because a function depends on an internal state doesn't mean that it is "impure". I don't think we should merge any of this.

**Comment by TaMaN2031A** [comment]
> I don't see the use of any of these comments. non-const class member functions will obviously depend on the internal state of the class. Most of these comments are not even true. Just because a function depends on an internal state doesn't mean that it is "impure". I don't think we should merge any of this.

Thank you for your time! I promise next pull requests will be touching this issue more precisely. Thank you again.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_env.h (hunk 6 above)
Comment (TaMaN2031A) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_env.h:52]: I was handling the concept with an academic [approach](https://www.syncfusion.com/blogs/post/pure-and-impure-functions-in-javascript-a-complete-guide.aspx#:~:text=An%20impure%20function%20is%20a,output%20for%20the%20same%20input.). A pure function is a function that returns the same value for the same input. In this class, the returned value "however value" in some functions depends on the state of the world_ variable.
The pull request as named was to just detect those functions, not enhance the code or put solutions to this at the current time. 
However, thank you for your comments it's really helpful!

Comment (henningkayser) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_env.h:52]: Depending on a class member doesn't make them impure. You don't seem to understand the term. The link points to some blog post, no academic approach visible. I'm closing this
